### PR TITLE
patron profile: add sort by due date

### PR DESCRIPTION
**Pull request by vgranata**
Add sort for due date in patron profile view.

Co-Authored-by: Valeria Granata <valeria@chaw.com>

## Why are you opening this PR?

Closes: https://github.com/rero/rero-ils/issues/2263

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

File: projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts [hunk 1]
```diff
@@ -65,31 +69,30 @@ export class PatronProfileLoansComponent implements OnInit, OnDestroy {
...
 
-  /** Initial records load */
-  private _initialLoad(): void {
-    this._loanQuery(1).subscribe((response: Record) => {
-      this._paginator.setRecordsCount(response.hits.total.value);
-      this.records = response.hits.hits;
+  /** Initial records load and sort */
+  private _initialLoadSorted(): void {
+    const pgr = this._paginator;
+    this._loanQueryAll().subscribe((response: Record) => {
```
Comment (Garfield-fr) [inline projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts:91]: It is not a good idea to load all the loans at once, because if there are a lot of them, it will make the thing explode. The pager is there to avoid this.

You really need to use the pager.

**Review by Garfield-fr — CHANGES_REQUESTED** [review]

File: projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.html [hunk 2]
```diff
@@ -18,38 +18,68 @@
...
     </div>
   </div>
 
+  <table class="table">
+  <tbody>
+    <ng-container *ngIf="!loaded; else display">
+      <div class="row">
+        <div class="col mt-2 font-weight-bold" translate>Loading in progress</div>
+      </div>
+    </ng-container>
```
Comment (zannkukai) [inline projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.html:61]: Seems strange to include a `div` as direct child of `tbody`.
`table` (+rows+colums) should be only included if loading is finished as before.

**Comment by zannkukai** [comment]
I continue about thinking at this problem during this weekend.
My personal conclusion is that paginator is maybe not a good idea on this page. As an user, I could be great to see all my current loans with one page without click on paginator links.

About technical it should be test with an account with a lot of loans (~100 loans) to check performance.
It should be discuss with other devs/PO members during a daily.

File: projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts [hunk 3]
```diff
@@ -29,14 +29,17 @@ import { PatronProfileMenuService } from '../patron-profile-menu.service';
 export class PatronProfileLoansComponent implements OnInit, OnDestroy {
 
   /** Observable subscription */
-  private _subscription = new Subscription();
+  _subscription: Subscription;
```
Comment (zannkukai) [inline projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts:32]: why did you remove `private` ?

File: projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts [hunk 4]
```diff
@@ -72,12 +77,12 @@ export class PatronProfileLoansComponent implements OnInit, OnDestroy {
     );
     this._subscription.add(
       this._patronProfileMenuService.onChange$.subscribe(() => {
-        this._initialLoad();
+        this.sortCriteria = 'duedate';
+        this.ngOnDestroy();
+        this.ngOnInit();
```
Comment (zannkukai) [inline projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts:82]: Arggh. Is it really a good idea to call life cycle hook function like that ? What's the feeling of @Garfield-fr and/or @jma.

File: projects/public-search/src/app/scss/styles.scss [hunk 5]
```diff
@@ -24,3 +24,17 @@
 .toast-container .ngx-toastr {
   width: 450px !important; // override width for all toastr message
 }
+.custom-select {
+  border: none;
+  background-color: transparent;
+  color: white;
+}
+select#sort-criteria.custom-select {
+  background-color:#343a40;
```
Comment (zannkukai) [inline projects/public-search/src/app/scss/styles.scss:33]: is it possible to use bootstrap variable here ? maybe `$table-dark-color` ?

File: projects/public-search/src/app/scss/styles.scss [hunk 6]
```diff
@@ -24,3 +24,17 @@
...
+}
+.input-group-text {
+  background-color:#343a40;
```
Comment (zannkukai) [inline projects/public-search/src/app/scss/styles.scss:36]: same here

File: projects/public-search/src/app/scss/styles.scss [hunk 7]
```diff
@@ -24,3 +24,17 @@
...
+  border: none;
+  padding: 1px;
```
Comment (zannkukai) [inline projects/public-search/src/app/scss/styles.scss:38]: use bootstrap function `map-get` if possible :: `map-get($spacers, 1)`

File: projects/public-search/src/app/scss/styles.scss (hunk 5 above)
Comment (vgranata) [reply projects/public-search/src/app/scss/styles.scss:33]: The changes in this file are to overwrite the style of the sort-list ng-core widget. 
I'm not sure how to use bootstrap in this case.

Comment (zannkukai) [reply projects/public-search/src/app/scss/styles.scss:33]: @vgranata

You can check https://github.com/rero/rero-ils-ui/blob/staging/projects/admin/src/app/acquisition/acquisition.scss (and some other files)
```scss
@import '~bootstrap/scss/_functions';
@import '~bootstrap/scss/_variables';
@import '/projects/admin/src/app/scss/variables';
```
allows to use bootstrap variables, bootstrap functions and project variables.
```scss
margin-bottom: map-get($spacers, 1);
```
you can use 1 (small gap space) to 5 (big gap space)
If you can't use a bootstrap variable for the background color ; define a project variable into `variables.scss` and use this variable in this file

**Review by zannkukai — COMMENTED** [review]
still some changes into SCSS file ;-)

File: projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts [hunk 8]
```diff
@@ -73,17 +87,10 @@ export class PatronProfileLoansComponent implements OnInit, OnDestroy {
...
     );
-    this._loanQuery(1).subscribe((response: Record) => {
-      this._initialLoad();
-    });
-  }
-
-  /** OnDestroy hook */
-  ngOnDestroy(): void {
-    this._subscription.unsubscribe();
   }
```
Comment (Garfield-fr) [inline projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts:94]: Do not delete this code

File: projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts [hunk 9]
```diff
@@ -103,6 +110,18 @@ export class PatronProfileLoansComponent implements OnInit, OnDestroy {
...
+  }
+
+  /**
+   * Allow to sort loans list using a sort criteria
+   * @param sortCriteria: the sort criteria to use for sorting the list
+   */
+  selectingSortCriteria(sortCriteria: string) {
+    this.sortCriteria = sortCriteria;
+    this._subscription.unsubscribe();
+    this._initialisePaginatorAndSubscription();
```
Comment (Garfield-fr) [inline projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts:123]: Use le current paginator and change only le value. Not necessary to reinitialize the paginator.

File: projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts [hunk 10]
```diff
@@ -103,6 +110,18 @@ export class PatronProfileLoansComponent implements OnInit, OnDestroy {
...
+      .getOnLoan(patronPid, page, this._paginator.getRecordsPerPage(), undefined, this.sortCriteria);
+  }
+
+  /**
+   * Allow to sort loans list using a sort criteria
+   * @param sortCriteria: the sort criteria to use for sorting the list
+   */
+  selectingSortCriteria(sortCriteria: string) {
+    this.sortCriteria = sortCriteria;
+    this._subscription.unsubscribe();
```
Comment (Garfield-fr) [inline projects/public-search/src/app/patron-profile/patron-profile-loans/patron-profile-loans.component.ts:122]: Remove this code and change with comment below

**Review by Garfield-fr — CHANGES_REQUESTED** [review]

**Review by Garfield-fr — APPROVED** [review]

**Comment by PascalRepond** [comment]
Works as intended. 

* [ ] Non-blocking: The UI of the sort column dropdown could be better. I would keep a similar look as the one in the pro UI. Keep the table heading color, add a fine border, and add the double arrow to the right : 
image

**Review by PascalRepond — APPROVED** [review]

**Review by PascalRepond — APPROVED** [review]
